spices/SPICE-0001-import-graph-analyzer-api.adoc

```diff
@@ -0,0 +1,336 @@
+
+  /// Mappings of modules from their in-language URI, to the URI that is used to fetch the module's contents.
+  external resolvedUris: Mapping<Uri, Uri>(keys == imports.keys)
+}
+----
+
+The `ImportAnalyzer`  is the in-language representation of the `pkl anaylze imports` CLI command.
+
+=== Java API
+
+A Java-only API is introduced to perform actual analysis.
+
+Unlike the CLI, this API only parses a module, and does not involve the evaluator.
+
+[source,java]
+----
+public interface ImportAnalyzer {
+    Results analyze(List<ModuleSource> moduleSources);
+
+    interface Results {
+        Map<URI, Set<URI>> getImports();
+
+        Map<URI, URI> getResolvedUris();
+    }
+}
+
+----
+
+=== Example
+
+This Pkl file:
+
+./path/to/my/foo.pkl
+[source,pkl]
+----
+import "@bar/Bar.pkl" // <1>
+import "package://example.com/bar@1.0.0#/Bar.pkl" // <2>
+import "pkl:reflect" // <3>
+import "..." // <4>
+import* "subdir/*.pkl" <5>
+
+res = if (foo) import("foo1.pkl") else import("foo2.pkl") // <6>
+----
+<1> local dependency import of package://example.com/bar@1.0.0
+<2> direct package import
+<3> stdlib import
+<4> triple-dot import (resolves to /path/to/foo.pkl)
+<5> globbed import (resolves to /path/to/my/subdir/foo.pkl)
+<6> conditional import; both are added to the dependency graph
+
+Produces the following result:
+
+[source,pkl]
+----
+imports {
+  ["projectpackage://example.com/bar@1.0.0#/Bar.pkl"] {} // <1>
+  ["file:///foo/bar/Bar.pkl"] {}
+  ["package://example.com/bar@1.0.0#/Bar.pkl"] {}
+  ["pkl:reflect"] {
+    "pkl:base"
+  }
+  ["pkl:base"] {
+    "pkl:jsonnet"
+    "pkl:xml"
+    "pkl:protobuf"
+  }
+  ["pkl:jsonnet"] {}
+  ["pkl:xml"] {}
+  ["pkl:protobuf"] {
+    "pkl:reflect"
+  }
+  ["file:///path/to/foo.pkl"] {}
+  ["file:///path/to/my/foo1.pkl"] {}
+  ["file:///path/to/my/foo2.pkl"] {}
+  ["file:///path/to/my/subdir/foo.pkl"] {}
+}
+resolvedUris {
+  ["projectpackage://example.com/bar@1.0.0#/Bar.pkl"] = "file:///path/to/local/bar/Bar.pkl" // <2>
+  ["file:///foo/bar/Bar.pkl"] = "file:///foo/bar/bar.pkl" // <3>
+  ["package://example.com/bar@1.0.0#/Bar.pkl"] = "package://example.com/bar@1.0.0#/Bar.pkl" // <4>
+  ["pkl:reflect"] = "pkl:reflect"
+  ["pkl:jsonnet"] = "pkl:jsonnet"
+  ["pkl:xml"] = "pkl:xml"
+  ["pkl:protobuf"] = "pkl:protobuf"
+  ["file:///path/to/foo.pkl"] = "file:///path/to/foo.pkl"
+  ["file:///path/to/my/foo1.pkl"] = "file:///path/to/my/foo1.pkl"
+  ["file:///path/to/my/foo2.pkl"] = "file:///path/to/my/foo2.pkl"
+  ["file:///path/to/my/subdir/foo.pkl"] = "file:///path/to/my/subdir/foo.pkl"
+}
+----
+<1> dependency notation turns into `projectpackage://`.
+<2> local dependency's resolve URI is a local file.
+<3> module URIs that don't resolve to anything further stay the same.
+<4> package URIs stay the same.
+
+=== Gradle Plugin Changes
+
+The input files for various Pkl Gradle tasks will switch to using the Java `ImportAnalyzer` to build its input files.
```

Comment:
How exactly will this work? Do you plan to run `ImportAnalyzer` at configuration time?

Reply:
I don’t know what the answer is, but this feels important to get right.

Reply:
The implementation is already in progress; we'll see.